packages/SwingSet/docs/garbage-collection.md

```diff
@@ -32,7 +30,76 @@ function makeStash() {
+
+```js
+function makeIncrementer() {
+  const counter = { count: 0, reads: 0 };
+  return harden({
+    increment() {
+      counter.count += 1;
+    },
+    read() {
+      counter.reads += 1;
+      return counter.count;
+    }
+  });
+}
+```
+
+For garbage-collection purposes, the incrementer can reach the counter. The platform must keep `counter` alive for at least as long as `incrementer` remain alive.
+
+## References
+
+Objects reference other objects in several ways:
+
+* Object properties
+* the Object's `.__prototype__`, which is effectively another property for GC purposes
+* code (including methods: object properties which are functions) which closes over variable bindings from an enclosing scope
+
+And of course most data structures (Arrays, Maps) hold strong references to their members.
+
+## Vat Roots
+
+Each vat has a set of "root references": these keep all other objects alive. These are the only starting points: if there is no pathway from a root reference to a target object, the target object is unreachable and will eventually be deleted.
+
+Vats have two kinds of root references. "Vat Globals" are top-level declarations (`const` and `let` declarations that appear in the highest scope of any module, outside any function definition). Top-level `const` declarations are eternal, while `let` declarations are mutable and could be added or removed later as the value is modified. These references are controlled entirely by the code that makes up a vat.
+
+"Vat Exports" are the subset of objects that have been used exported through the kernel to other vats. Such objects are added to the export table when they are used as an argument in an outbound message, or in the resolution of a Promise that some other vat is following. Each exported object is allocated a new integer and assigned a "vref" (vat-reference ID), in the form `o+NN`. A special initial "root object" is defined when the vat is first constructed (the return object of `buildRootObject()`) and assigned vref `o+0`.
+
+Objects are removed from the export table when no other vat retains a reference. If an object is not exported, and not a global, and not transitively reachable by any object in either of those two categories, then the object is unreachable and will be deleted.
+
+![Vat Roots diagram](./images/gc/vat-roots.png "Vat Roots")
+
+In this example, A is the "root object", A and B are vat exports, and C is a Vat Global. D is kept alive by virtue of the reference from C. If C were modified and dropped its reference to D, D would be deleted.
+
+## Weak References
+
+JavaScript, like many languages, offers the notion of a "Weak Reference". In JS, this is exposed in the `WeakRef` object. A weak reference can be used to reach the target *if* it is still around, but does not keep its target alive on its own.
+
+![WeakRef diagram](./images/gc/weakref.png "WeakRef")
```

Comment:
Nit: Illo is slightly confusing because the caption talks about `wr.deref() === a`, which looks like JS code, but of course if such code were there it would constitute a strong reference to A and hold it in memory.

Reply:
Hm, do you think pseudocode like `wr.deref() -> a` would be better? or `x = wr.deref(); // x === a` ? I don't want to let it grow too wide, so I can still fit three examples next to each other in the following diagram, but I totally agree that legal JS code is a confusing way to express the "is it reachable or not" question.

Reply:
I think I like the `->` better, but go with what works best in terms of layout.

Reply:
I'll try `wr.deref(); // a` and `wr.deref(); // undefined`.